c7n/filters/core.py

```diff
@@ -816,3 +818,107 @@ def get_resource_value(self, resource):
         if capture is None:  # regex didn't capture anything
             return None
         return capture.group(1)
+
+
+class ReduceFilter(ValueFilter):
+    """Generic reduce filter to group, sort, and limit your resources.
+
+    This example will select the longest running instance from each ASG,
+    hen randomly choose 10% of those, maxing at 15 total instances.
+
+    filters:
+      - "tag:aws:autoscaling:groupName": present
+      - type: reduce
+        group_by: "tag:aws:autoscaling:groupName"
+        sort_by: "LaunchTime"
+        order: asc
+        limit: 1
+      - type: reduce
+        order: randomize
+        limit: 15
+        limit-percent: 10
+    """
+    annotate = False
+
+    schema = {
+        'type': 'object',
+        # Doesn't mix well with inherits that extend
+        'additionalProperties': False,
+        'required': ['type'],
+        'properties': {
+            # Doesn't mix well as enum with inherits that extend
+            'type': {'enum': ['reduce']},
+            'group_by': {'type': 'string'},
+            'sort_by': {'type': 'string'},
```

Comment:
so any reason on the different hyphenation within the same filter? re sort_by limit-percent?

Reply:
heh... started with underscore (not sure why) ... then later added limit-percent to maintain consistency with it's use elsewhere.  i'm changing them to group-by and sort-by...